session: allow log configuration using Logger interface.

Kulezi opened the pull request:
This PR provides a Logger interface allowing to supply
a custom logger as part of (Session)ConnConfig.

It also provides three implementations of Logger:
- DefaultLogger, logging only warnings
- DebugLogger, logging everything that can be useful for debugging,
  it is used by default in tests
- NopLogger, logging nothing

Fixes #271

zimnx commented on transport/logger.go:
```diff
@@ -0,0 +1,69 @@
+package transport
```
how does logger relates to transport? It feels like it belongs to root package

Kulezi replied:
Almost all logging takes place in the transport package, and putting the entire logger inside of scylla package creates an import cycle, transport package should not have dependencies in scylla package.

I suppose Logger interface could be declared identically in both transport and scylla packages instead of aliasing it in scylla package so it shows up as the whole declaration instead of an alias in docs.

It also could have its own package but that seems excessive to me.

piodul commented on transport/conn.go:
```diff
@@ -383,6 +383,10 @@ func OpenShardConn(ctx context.Context, addr string, si ShardInfo, cfg ConnConfi
+		if err := ctx.Err(); err != nil {
+			return nil, fmt.Errorf("aborted opening connection on shard %d: %w", si.Shard, err)
+		}
+
```
Does the second commit (`transport: conn, abort OpenShardConn faster when context is done`) belong to this PR? It looks like it can be submitted separately and reviewed independently.

Kulezi replied:
Moved to #286.

zimnx commented on logger/logger.go:
```diff
@@ -0,0 +1,69 @@
+package logger
```
```suggestion
package log
```
"log" is in sync with other logging packages

zimnx commented on session.go:
```diff
@@ -66,6 +66,18 @@ var (
 	Lz4    Compression = frame.Lz4
 )
 
+func NewDefaultLogger() log.Logger {
+	return log.NewDefaultLogger()
+}
+
+func NewDebugLogger() log.Logger {
+	return log.NewDebugLogger()
+}
+
+func NewNopLogger() log.Logger {
+	return log.NopLogger{}
+}
+
```
These shouldn't be exposed in root package, users aren't generally interested such details, default session should use default logger. Advanced users who would like to change logging details, can import log package.

Kulezi commented:
Rebased, renamed the new logging package to log. Integration tests don't fail anymore after #283 was merged. 

Please let me know if I can merge this @zimnx.

zimnx reviewed (approved):
LGTM, thanks
